### ISSUE TITLE
PoisonError encountered

### ISSUE THREAD

**Issue by yoshuawuyts** [nnethercote/dhat-rs#14]
After having run into https://github.com/nnethercote/dhat-rs/issues/13 on MacOS, we ran `dhat` on Linux. For simple examples this seemed to work well. But once we tried a load test we managed to trigger a `PoisonError`:

```txt
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/dhat-0.2.2/src/lib.rs:247:70
```

We have no idea what is causing this error, but we were hoping that filing it might be helpful. Thanks!

## System Details
- OS: Linux (Not sure which flavor; but it's a recent install)
- Rust version: 1.49

## Source of Panic

The `unwrap` call on line 247 seems to be the root of the issue.

https://github.com/nnethercote/dhat-rs/blob/bfbd533b1be15e48842d8aae98935cd80a53aeb7/src/lib.rs#L244-L248

## Stack Trace
<details>
<pre>
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/dhat-0.2.2/src/lib.rs:247:70
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   1: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:1268:5
   3: core::result::Result<T,E>::unwrap
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:973:23
   4: <dhat::DhatAlloc as core::alloc::global::GlobalAlloc>::alloc::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/dhat-0.2.2/src/lib.rs:247:51
   5: dhat::if_ignoring_allocs_else
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/dhat-0.2.2/src/lib.rs:620:9
   6: <dhat::DhatAlloc as core::alloc::global::GlobalAlloc>::alloc
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/dhat-0.2.2/src/lib.rs:244:9
   7: __rg_alloc
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:30:1
   8: alloc::alloc::alloc
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:84:14
   9: alloc::alloc::Global::alloc_impl
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:164:73
  10: <alloc::alloc::Global as core::alloc::AllocRef>::alloc
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/alloc.rs:224:9
  11: alloc::raw_vec::finish_grow
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:495:9
  12: alloc::raw_vec::RawVec<T,A>::grow_amortized
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:428:19
  13: alloc::raw_vec::RawVec<T,A>::try_reserve
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:317:13
  14: alloc::raw_vec::RawVec<T,A>::reserve
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/raw_vec.rs:311:24
  15: alloc::vec::Vec<T>::reserve
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:505:9
  16: alloc::vec::Vec<T>::push
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:1210:13
  17: regex_syntax::ast::parse::ParserI<P>::parse_with_comments
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-syntax-0.6.22/src/ast/parse.rs:1007:22
  18: regex_syntax::ast::parse::ParserI<P>::parse
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-syntax-0.6.22/src/ast/parse.rs:964:9
  19: regex_syntax::ast::parse::Parser::parse
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-syntax-0.6.22/src/ast/parse.rs:338:9
  20: regex_syntax::parser::Parser::parse
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-syntax-0.6.22/src/parser.rs:196:19
  21: regex::exec::ExecBuilder::parse
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-1.4.3/src/exec.rs:242:17
  22: regex::exec::ExecBuilder::build
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-1.4.3/src/exec.rs:307:22
  23: regex::re_builder::unicode::RegexBuilder::build
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-1.4.3/src/re_builder.rs:70:21
  24: regex::re_unicode::Regex::new
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-1.4.3/src/re_unicode.rs:176:9
  25: <<tracing_subscriber::filter::env::directive::Directive as core::str::traits::FromStr>::from_str::DIRECTIVE_RE as core::ops::deref::Deref>::deref::__static_ref_initialize
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/directive.rs:179:46
  26: core::ops::function::FnOnce::call_once
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  27: lazy_static::lazy::Lazy<T>::get::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-1.4.0/src/inline_lazy.rs:31:29
  28: std::sync::once::Once::call_once::{{closure}}
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/once.rs:261:41
  29: std::sync::once::Once::call_inner
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sync/once.rs:419:21
  30: std::sync::once::Once::call_once
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sync/once.rs:261:9
  31: lazy_static::lazy::Lazy<T>::get
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-1.4.0/src/inline_lazy.rs:30:9
  32: <<tracing_subscriber::filter::env::directive::Directive as core::str::traits::FromStr>::from_str::DIRECTIVE_RE as core::ops::deref::Deref>::deref::__stability
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-1.4.0/src/lib.rs:142:21
  33: <<tracing_subscriber::filter::env::directive::Directive as core::str::traits::FromStr>::from_str::DIRECTIVE_RE as core::ops::deref::Deref>::deref
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-1.4.0/src/lib.rs:144:17
  34: <tracing_subscriber::filter::env::directive::Directive as core::str::traits::FromStr>::from_str
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/directive.rs:210:20
  35: core::str::<impl str>::parse
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/str/mod.rs:2209:9
  36: tracing_subscriber::filter::env::EnvFilter::new::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/mod.rs:156:72
  37: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:269:13
  38: core::iter::traits::iterator::Iterator::find_map::check::{{closure}}
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:2257:32
  39: core::iter::traits::iterator::Iterator::try_fold
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1888:21
  40: core::iter::traits::iterator::Iterator::find_map
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:2263:9
  41: <core::iter::adapters::FilterMap<I,F> as core::iter::traits::iterator::Iterator>::next
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/mod.rs:1245:9
  42: <alloc::vec::Vec<T> as alloc::vec::SpecFromIterNested<T,I>>::from_iter
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2082:32
  43: <alloc::vec::Vec<T> as alloc::vec::SpecFromIter<T,I>>::from_iter
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:2119:9
  44: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec.rs:1959:9
  45: core::iter::traits::iterator::Iterator::collect
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:1670:9
  46: tracing_subscriber::filter::env::EnvFilter::from_directives
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/mod.rs:253:34
  47: tracing_subscriber::filter::env::EnvFilter::new
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/mod.rs:163:9
  48: core::ops::function::FnOnce::call_once
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  49: core::result::Result<T,E>::map
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:508:25
  50: tracing_subscriber::filter::env::EnvFilter::from_env
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/mod.rs:150:9
  51: tracing_subscriber::filter::env::EnvFilter::from_default_env
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.15/src/filter/env/mod.rs:144:9
  52: query_engine::init_logger
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:69:18
  53: query_engine::main::main::{{closure}}::main::{{closure}}
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:52:9
  54: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  55: query_engine::main::main::{{closure}}
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:42:12
  56: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  57: query_engine::main::{{closure}}
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:40:1
  58: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  59: <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/builder.rs:199:17
  60: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/task_locals_wrapper.rs:60:13
  61: std::thread::local::LocalKey<T>::try_with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  62: std::thread::local::LocalKey<T>::with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  63: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/task_locals_wrapper.rs:55:9
  64: <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/builder.rs:197:13
  65: <futures_lite::future::Or<F1,F2> as core::future::future::Future>::poll
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-lite-1.11.3/src/future.rs:526:33
  66: async_executor::Executor::run::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.4.0/src/lib.rs:236:9
  67: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  68: async_executor::LocalExecutor::run::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-1.4.0/src/lib.rs:437:9
  69: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  70: async_io::driver::block_on
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-io-1.3.1/src/driver.rs:142:33
  71: async_global_executor::reactor::block_on::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/reactor.rs:3:18
  72: tokio::runtime::context::enter
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.24/src/runtime/context.rs:72:5
  73: tokio::runtime::handle::Handle::enter
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.24/src/runtime/handle.rs:76:9
  74: async_global_executor::tokio02::enter
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/tokio02.rs:5:5
  75: async_global_executor::reactor::block_on::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/reactor.rs:9:18
  76: async_global_executor::reactor::block_on
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/reactor.rs:12:5
  77: async_global_executor::executor::block_on::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/executor.rs:26:36
  78: std::thread::local::LocalKey<T>::try_with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  79: std::thread::local::LocalKey<T>::with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  80: async_global_executor::executor::block_on
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/executor.rs:26:5
  81: async_std::task::builder::Builder::blocking::{{closure}}::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/builder.rs:171:25
  82: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/task_locals_wrapper.rs:60:13
  83: std::thread::local::LocalKey<T>::try_with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  84: std::thread::local::LocalKey<T>::with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  85: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/task_locals_wrapper.rs:55:9
  86: async_std::task::builder::Builder::blocking::{{closure}}
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/builder.rs:168:17
  87: std::thread::local::LocalKey<T>::try_with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  88: std::thread::local::LocalKey<T>::with
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  89: async_std::task::builder::Builder::blocking
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/builder.rs:161:9
  90: async_std::task::block_on::block_on
             at /home/dom/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.9.0/src/task/block_on.rs:33:5
  91: query_engine::main
             at /home/dom/Documents/Repositories/prisma-engines/query-engine/query-engine/src/main.rs:40:1
  92: core::ops::function::FnOnce::call_once
             at /home/dom/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
</pre>
</details>

**Comment by repi** [comment]
getting this exact problem also when testing this in our quite threaded app on Windows. Was a double panic:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\repi\.cargo\registry\src\github.com-1ecc6299db9ec823\dhat-0.2.2\src\lib.rs:308:67
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', C:\Users\repi\.cargo\registry\src\github.com-1ecc6299db9ec823\dhat-0.2.2\src\lib.rs:334:70
```

the first panic (that probably caused the second poison panic) was the latest unwrap here:

```rust
                    // Remove the record of the existing live block and get the
                    // `PpInfo` (which must be present, because we created it
                    // if necessary when this block was allocated.)
                    let h = g.heap.as_mut().unwrap();
                    let LiveBlock {
                        pp_info_idx,
                        allocation_instant: _,
                    } = h.live_blocks.remove(&(old_ptr as usize)).unwrap();
```

any ideas on this @nnethercote? thx!

**Comment by nnethercote** [comment]
Hi! Sorry for the lack of response for such a long time on this issue, I wasn't able to work on this for a while. But I'm back in a position where I can. Hopefully I'll look at this soon.

**Comment by nnethercote** [comment]
@repi: your problem is the same as the one reported in #13, which I will fix shortly.

@yoshuawuyts: in @repi's report there was a double panic. I saw a similar thing when testing #13, but I overlooked it at first because in my terminal the first panic only printed one or two lines, and then the second panic (which was a `PoisonError`) printed lots of lines. So I wonder if you might have overlooked an initial panic as well?

I've been so slow responding to this that you probably don't remember and don't have the relevant output saved. So I will close this issue, because it's probably impossible to action. (If you do have any extra info, I'd be happy to hear about it.) I am currently improving the crate's testing, so hopefully this kind of crash won't happen any more once that's done.
